garage/tf/models/base.py

```diff
@@ -0,0 +1,313 @@
+        raise NotImplementedError
+
+    def __getstate__(self):
+        """Object.__getstate__."""
+        args, kwargs = self.init_spec
+        return {'args': args, 'kwargs': kwargs, 'parameters': self.parameters}
+
+    def __setstate__(self, data):
+        """Object.__setstate__."""
+        # TODO(krzentner): There has got to be a better way.
+        out = type(self)(*data['args'], **data['kwargs'])
+        out._default_parameters = data['parameters']
+        self.__dict__.update(out.__dict__)
+
+
+class TfModel(AutoPickableModel):
+    r"""
+    Model class for TensorFlow.
+
+    This class is a parent-child design.
+    The parent model contains the parameters, i.e. the structure, and
+    two child model are constructed with input_1 and input_2. Inputs and
+    outputs of the parent model will be binded with the default child, which
+    is the first child constructed.
+
+    When new child model is created, it reuses the parameter from the
+    parent model and can be accessed by calling model.child_2, where
+    child_2 is the name given when building this particular child model.
+    If a child is constructed without given a name, the name "default" will
+    be used.
+
+    The design is illustrated as the following:
+
+         input_1                      input_2
+            |                            |
+    ============== Parent (TfModel)===============
+    |       |                            |       |
+    |       |          Parameters        |       |
+    |    ===========  /           \  =========== |
+    |    | Child 1 | /             \ | Child 2 | |
+    |    |(TfModel)|/               \|(TfModel)| |
+    |    ===========                 =========== |
+    |       |                            |       |
+    ==============================================
+            |                            |
+       (model.outputs) OR                |
+    (model.default.outputs)        model.child_2.outputs
+
+    The parameters inside a model will be initialized when calling build().
+    Do not call tf.global_variable_initializers() after building a model as it
+    will reassign random weights to the model.
+
+    Examples are also available in tests/garage/tf/models/test_new_model.
+
+    Args:
+      name (str): Name of the model. It will also become the variable scope
+      of the model. Every model should have a unique name.
+    """
+
+    def __init__(self, name):
+        super().__init__()
+        self._name = name or self._name
+        self._is_parent = True
+        self._is_built = False
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
+
+    @classmethod
+    def _new_child(cls, other, name):
+        """Class method for constructing child model with the given name."""
+        twin = cls(**other.init_spec[1])
+        twin._name = name
+        twin._is_parent = False
```

Comment:
again it's important to get the terminology right.

parent/child implies a tree structure.

Reply:
i think you will need different names